### PR TITLE
🤖 Automatically set user status to ⛔ DND outside of  📅 availability

### LINKED ISSUE #7321: Automatic status handling based on availability
The goal is to add an option to `index.php/settings/user/groupware`
The checkbox would basically mean that when ever the user's "start" is reached DND is removed when it was set automatically.
At "end" the status is set to DND when the status is not manually set to DND already.

Similar to the automatic call status a "Backup status" should be created which would be restored on the next morning.

The setting implementation is in the DAV app (and should only be visible when user_status is enabled).
The Background job changing the status should live in the user status app

### PR DESCRIPTION
**Pull request by nickvergessen**
- [x] Add a background job that changes the status automatically to DND and restore the previous status again
- [x] Add option in the personal settings to opt-in on the automation

Fix https://github.com/nextcloud/spreed/issues/7321

### CODE REVIEW

File: apps/dav/lib/BackgroundJob/UserStatusAutomation.php [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+		$query->executeStatement();
+
+		$this->logger->debug('Updated user status automation last_run to ' . $timestamp . ' for user ' . $userId);
+	}
+
+	/**
+	 * @param string $userId
+	 * @return false|string
+	 */
+	protected function getAvailabilityFromPropertiesTable(string $userId) {
```
Comment (tcitworld) [inline apps/dav/lib/BackgroundJob/UserStatusAutomation.php:171]: You could also do this without hooking directly into the database through the use of the dav server:

There's an example here:
https://github.com/nextcloud/server/blob/827e622b50e26fa583d3df65d52485850a039e2e/apps/dav/lib/CalDAV/Schedule/Plugin.php#L500-L520

Comment (nickvergessen) [reply apps/dav/lib/BackgroundJob/UserStatusAutomation.php:171]: So the issue is that this is all not injectable at this point. I would need to set up all the objects similarly to:
https://github.com/nextcloud/server/blob/827e622b50e26fa583d3df65d52485850a039e2e/apps/dav/appinfo/v1/caldav.php#L96
Only then I have the $server object to be able to register the plugin and then run the propfind?

Doesn't look like something I want to touch at this point :/

**Review by miaulalala — APPROVED** [review]
Approved :shipit:

File: apps/dav/lib/BackgroundJob/UserStatusAutomation.php (hunk 1 above)
Comment (ChristophWurst) [reply apps/dav/lib/BackgroundJob/UserStatusAutomation.php:171]: Noted :notebook: :pencil2:

File: apps/dav/src/views/Availability.vue [hunk 2]
```diff
@@ -24,6 +24,10 @@
 			:l10n-saturday="$t('dav', 'Saturday')"
 			:l10n-sunday="$t('dav', 'Sunday')" />
 
+		<CheckboxRadioSwitch :checked.sync="automated">
+			{{ $t('dav', 'Automatically set user status to "Do not distrub" outside of visibility to mute all notifications.') }}
```
Comment (tcitworld) [inline apps/dav/src/views/Availability.vue:28]: I'm not sure this is easily understandable (what's `visibility` ?) by most people. Maybe rephrase and/or add a explanation underneath?

**Comment by nickvergessen** [comment]
Rebased to fix the import conflict on  `apps/dav/lib/AppInfo/Application.php`

cc @ChristophWurst can someone from groupware or you yourself help with the review?

**Comment by nickvergessen** [comment]
Rebased again to solve conflicts

File: apps/dav/lib/Settings/AvailabilitySettings.php [hunk 3]
```diff
@@ -27,10 +27,35 @@
...
 
 class AvailabilitySettings implements ISettings {
+	protected IUserSession $userSession;
+	protected IConfig $config;
+	protected IInitialState $initialState;
+
+	public function __construct(IUserSession $userSession,
+								IConfig $config,
+								IInitialState $initialState) {
+		$this->userSession = $userSession;
```
Comment (CarlSchwan) [inline apps/dav/lib/Settings/AvailabilitySettings.php:43]: maybe directly dependency inject `?string $userId` ?

**Review by CarlSchwan — APPROVED** [review]
